src/Domain/LeanCode.DomainModels.MassTransitRelay/Middleware/EventsPublisherMiddleware.cs

```diff
@@ -0,0 +1,58 @@
+using LeanCode.DomainModels.Model;
+using LeanCode.Pipelines;
+using MassTransit;
+using Microsoft.AspNetCore.Http;
+
+namespace LeanCode.DomainModels.MassTransitRelay.Middleware;
+
+public class EventsPublisherMiddleware
+{
+    private readonly Serilog.ILogger logger = Serilog.Log.ForContext<EventsPublisherMiddleware>();
+    private readonly RequestDelegate next;
+    private readonly AsyncEventsInterceptor interceptor;
+
+    public EventsPublisherMiddleware(RequestDelegate next, AsyncEventsInterceptor interceptor)
+    {
+        this.next = next;
+        this.interceptor = interceptor;
+    }
+
+    public async Task InvokeAsync(HttpContext httpContext, IPublishEndpoint publishEndpoint)
+    {
+        var events = await interceptor.CaptureEventsOfAsync(() => next(httpContext));
+
+        if (events.Count > 0)
+        {
+            await PublishEventsAsync(publishEndpoint, events, httpContext.RequestAborted);
+        }
+    }
+
+    private Task PublishEventsAsync(IPublishEndpoint publishEndpoint, List<IDomainEvent> events, CancellationToken ct)
+    {
+        logger.Debug("Publishing {Count} raised events", events.Count);
+        var conversationId = Guid.NewGuid();
+
+        var publishTasks = events.Select(evt => PublishEventAsync(publishEndpoint, evt, conversationId, ct));
```

Comment:
Can't we use some `PublishBatch` here?

Reply:
If this "just" appends to an outbox (so - dbcontext), is it possible to switch to sync?

Reply:
No, there're no sync counterparts.
Digging through MT onion of abstraction I see that indeed it's 'just' adding to db context, but with `AddAsync` :confused: